apis/oam/v1alpha2/core_types.go

```diff
@@ -0,0 +1,375 @@
+	Traits []ComponentTrait `json:"traits,omitempty"`
+
+	// Scopes in which the specified component should exist.
+	// +optional
+	Scopes []ComponentScope `json:"scopes,omitempty"`
+}
+
+// An ApplicationConfigurationSpec defines the desired state of a
+// ApplicationConfiguration.
+type ApplicationConfigurationSpec struct {
+	// Components of which this ApplicationConfiguration consists. Each
+	// component will be used to instantiate a workload.
+	Components []ApplicationConfigurationComponent `json:"components"`
+}
+
+// A WorkloadReference refers to an OAM workload resource.
+type WorkloadReference struct {
+	// APIVersion of the referenced workload.
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of the referenced workload.
+	Kind string `json:"kind"`
+
+	// Name of the referenced workload.
+	Name string `json:"name"`
+
+	// UID of the referenced workload.
+	// +optional
+	UID *types.UID `json:"uid,omitempty"`
+}
+
+// A TraitReference refers to an OAM workload trait.
+type TraitReference struct {
+	// APIVersion of the referenced workload.
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of the referenced workload.
+	Kind string `json:"kind"`
+
+	// Name of the referenced workload.
+	Name string `json:"name"`
+
+	// UID of the referenced workload.
+	// +optional
+	UID *types.UID `json:"uid,omitempty"`
+}
+
+// A TraitStatus represents the state of a trait.
+type TraitStatus string
+
+// Trait statuses.
+const (
+	TraitStatusUnknown = "Unknown"
+	TraitStatusError   = "Error"
+	TraitStatusCreated = "Created"
+)
+
+// A WorkloadTrait represents a trait associated with a workload.
+type WorkloadTrait struct {
+	// Reference to a trait created by an ApplicationConfiguration.
+	Reference TraitReference `json:"traitRef"`
+
+	// Status of this trait.
+	// +kubebuilder:validation:Enum=Unknown;Error;Created
+	Status TraitStatus `json:"status,omitempty"`
+}
+
+// A WorkloadStatus represents the current state of a workload.
+type WorkloadStatus string
+
+// Workload statuses.
+const (
+	WorkloadStatusUnknown = "Unknown"
+	WorkloadStatusError   = "Error"
+	WorkloadStatusCreated = "Created"
+)
+
+// An ApplicationConfigurationWorkload represents a workload associated with an
+// ApplicationConfiguration.
+type ApplicationConfigurationWorkload struct {
+	// Reference to a workload created by an ApplicationConfiguration.
+	Reference WorkloadReference `json:"workloadRef,omitempty"`
+
+	// Traits associated with this workload.
+	Traits []WorkloadTrait `json:"traits,omitempty"`
+
+	// Status of this workload.
+	// +kubebuilder:validation:Enum=Unknown;Error;Created
+	Status WorkloadStatus `json:"status,omitempty"`
+}
+
+// An ApplicationConfigurationStatus represents the observed state of a
+// ApplicationConfiguration.
+type ApplicationConfigurationStatus struct {
+	runtimev1alpha1.ConditionedStatus `json:",inline"`
+
+	// Workloads created by this ApplicationConfiguration.
+	Workloads []ApplicationConfigurationWorkload `json:"workloads,omitempty"`
```

Comment:
Are they workloads or components?

Reply:
This is an array of the CRs (like `ContainerizedWorkload`) that were rendered by the AppConfig. I believe we should call these workloads. This way the mental model is "an AppConfig uses components to produce workloads", rather than "an AppConfig uses components to produce components".